src/Microsoft.ML.FastTree/GamModelParameters.cs

```diff
@@ -0,0 +1,911 @@
+        }
+
+        private void Map(in VBuffer<float> features, ref float response)
+        {
+            Host.CheckParam(features.Length == _inputLength, nameof(features), "Bad length of input");
+
+            double value = Intercept;
+            var featuresValues = features.GetValues();
+
+            if (features.IsDense)
+            {
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(i, out int j))
+                        value += GetBinEffect(j, featuresValues[i]);
+                }
+            }
+            else
+            {
+                var featuresIndices = features.GetIndices();
+                // Add in the precomputed results for all features
+                value += _valueAtAllZero;
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(featuresIndices[i], out int j))
+                        // Add the value and subtract the value at zero that was previously accounted for
+                        value += GetBinEffect(j, featuresValues[i]) - GetBinEffect(j, 0);
+                }
+            }
+
+            response = (float)value;
+        }
+
+        internal double GetFeatureBinsAndScore(in VBuffer<float> features, int[] bins)
+        {
+            Host.CheckParam(features.Length == _inputLength, nameof(features));
+            Host.CheckParam(Utils.Size(bins) == _numFeatures, nameof(bins));
+
+            double value = Intercept;
+            var featuresValues = features.GetValues();
+            if (features.IsDense)
+            {
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(i, out int j))
+                        value += GetBinEffect(j, featuresValues[i], out bins[j]);
+                }
+            }
+            else
+            {
+                var featuresIndices = features.GetIndices();
+                // Add in the precomputed results for all features
+                value += _valueAtAllZero;
+                Array.Copy(_binsAtAllZero, bins, _numFeatures);
+
+                // Update the results for features we have
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(featuresIndices[i], out int j))
+                        // Add the value and subtract the value at zero that was previously accounted for
+                        value += GetBinEffect(j, featuresValues[i], out bins[j]) - GetBinEffect(j, 0);
+                }
+            }
+            return value;
+        }
+
+        private double GetBinEffect(int featureIndex, double featureValue)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
```

Comment:
>Contracts [](start = 12, length = 9)

Please change `Contracts` to `Host` where it possible.

Reply:
Done.